### DISCUSSION TITLE
how error messages are handled in dbus-cxx ?

### DISCUSSION THREAD

**Discussion by lakabd** [dbus-cxx/dbus-cxx#142] [Q&A] [answered]
I'm big newbie in dbus programming.
Currently I'm working on upgrading an application that uses dbus-cplusplus bindings to use dbus-cxx. So in my journey of understanding how dbus-cxx works, I'm struggling to find out how errors are managed: 
Let's say a client wants to call a method of a dbus server, in this method a check is performed and if it fails it shall return an error to the client. I know that a DBus::ErrorMessage can be used here but I'm not sure how to !!
I'm using an introspection file (.xml) to generate the adapter code, but I cannot figure out if a I need to add the instantiation of DBus::ErrorMessage/DBus:CallMessage by myself ? what is the right way to handle the return of error messages from adapter methods ?

Any help is greatly appreciated.

Thank you
Abderrahim

**Comment by rm5248** [comment]
There's currently no easy way to send back a specific error message.

If a method is being called, an error message will be returned if an exception is thrown.  Generally most DBus APIs that I have seen do not return errors back, they are much more like C APIs where the return value determines if the call was successful or not.

If you need to send back a specific error message, it should be pretty simple to add a new `catch` clause to the statement in `methodbase.h` [here](https://github.com/dbus-cxx/dbus-cxx/blob/ca478ce28d70adffb76287a584abcabacd3ebc18/dbus-cxx/methodbase.h#L115) that would catch the `DBus::Error` being thrown.

**Comment by lakabd** [comment]
Hi Robert,
Thank you for your answer.

I have created the example code below for a hellowrold adapter, and I'm testing it via the `busctl` command.
However, When I added a throw in my method definition `say_hello`, the bustctl does not receive any message (it just hangs until it timeouts out) !!
```
#include <dbus-cxx.h>
#include <glibdispatcher.h>
#include <iostream>
#include <unistd.h>
#include "HelloWorldAdapter.h"

#define CONNECTION_NAME "org.example.helloworld"
#define INTERFACE_NAME "HelloWorld.iface"
#define PATH_NAME "/org/example/helloworld"

class HelloWorld : public org_example_helloworld {
    public:
        std::string say_hello() override {
            if(1)
                throw std::runtime_error("error");
            else
                return "Hello World !";
        }
};

int main() {
    auto main_loop = g_main_loop_new(NULL, FALSE);
    std::shared_ptr<DBus::Dispatcher> dispatcher = DBus::GLib::GLibDispatcher::create();
    std::shared_ptr<DBus::Connection> conn = dispatcher->create_connection(DBus::BusType::SESSION);
    if( conn->request_name(CONNECTION_NAME, DBUSCXX_NAME_FLAG_REPLACE_EXISTING ) != DBus::RequestNameResponse::PrimaryOwner )
        return 1;
    HelloWorld* pointer = new HelloWorld();
    std::shared_ptr<org_example_helloworldInterface> interface = org_example_helloworldInterface::create(pointer, INTERFACE_NAME);
    std::shared_ptr<HelloWorldAdapter> adapter = HelloWorldAdapter::create(conn, interface, PATH_NAME, DBus::ThreadForCalling::DispatcherThread);
    g_main_loop_run(main_loop);
    return 0;
}
```
```
<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
<node cppname="HelloWorld">
  <interface name="org.example.helloworld">
    <method name="say_hello">
      <arg name="data" type="s" direction="out" />
    </method>
  </interface>
</node>
```
I'm calling busctl as follows:
`$ busctl --user call org.example.helloworld /org/example/helloworld HelloWorld.iface say_hello`

do you know what is going wrong here ?

**Reply by rm5248** [reply]
I have not used busctl before, but possibly it is not handling the error return.  You might want to try with `dbus-send` and see if that works(make sure to use the `--print-reply` flag to get the reply).

One thing that may be useful as well is to run `dbus-monitor` in another terminal and using that to decode the messages that go between applications.

**Reply by lakabd** [reply]
I have tried with `dbus-send` but with no success (even with --print-reply). Using `dbus-monitor` I cannot see any error message on the bus when running the dbus-send command (by comparing the output with and without the `throw`call in my `say_hello`method).
However, when I enabled dbus-cxx debug traces, I was able to see that SendmsgTransport sends the following message when I have the throw in say_hello:
```
0x4352F700 DBus.priv.SendmsgTransport [TRACE] - Going to send the following bytes: 
000000: 42 03 00 01 00 00 00 05   00 00 00 03 00 00 00 3f   B.......  .......?
000010: 04 01 73 00 00 00 00 21   6f 72 67 2e 66 72 65 65   ..s....!  org.free
000020: 64 65 73 6b 74 6f 70 2e   44 42 75 73 2e 45 72 72   desktop.  DBus.Err
000030: 6f 72 2e 46 61 69 6c 65   64 00 00 00 00 00 00 00   or.Faile  d.......
000040: 05 01 75 00 00 00 00 02   08 01 67 00 01 73 00 00   ..u.....  ..g..s..
000050: 00 00 00 00 00   
```  
But I don't know why neither `dbus-monitor` nor `dbus-send` are seeing the message, I tried on another computer but the same again (btw am using debian 10). I have tried also to use the StandAlone dispatcher instead of the glib one but this changes nothing.

Here is the command I used for dbus-send:
``` 
dbus-send --session --dest=org.example.helloworld --type=method_call  --print-reply /org/example/helloworld HelloWorld.iface.say_hello
```

**Reply by rm5248** [reply]
That sounds like there could be a bug, I can take a look at it later.

**Reply by rm5248** [reply, marked as answer]
It looks like there is a problem.  See #143 for the bug.

**Reply by lakabd** [reply]
Hi Robert,

Indeed, It looked very odd. Thank you very much for the fix !

Best regards,
Abderrahim
